_posts/function-contracts.md

```diff
@@ -0,0 +1,298 @@
+# Function Contracts for Kani
+
+I this blogpost we discuss a new feature we’re developing for Kani: Function Contracts. It’s now available as an unstable feature with the `-Zfunction-contracts` flag. If you would like to learn more about the development and implementation details of this feature and leave feedback please refer to [the RFC](https://model-checking.github.io/kani/rfc/rfcs/0009-function-contracts.html).
+
+## Introduction
+
+Verification is a costly art. If you are an active user of Kani you know that the runtime of a proof does not scale linearly with the amount of code under verification and you probably have a number of functions and ideas for harnesses you’d like Kani to verify, but it just takes prohibitively long.
+
+We [previously](https://model-checking.github.io/kani-verifier-blog/2023/02/28/kani-internship-projects-2022-stubbing.html) blogged about stubbing, a technique where a function that performs a costly computation, for instance trough recursion or a long running loop, could be replaced with a function that is cheaper to execute in the verifier. This replacement approximates the behavior of the function it replaces and/or uses techniques only available in a verifier, like non-deterministic values (`kani::any`). By successively replacing costly functions with stubs Kani can check ever larger proofs but it also comes with a downside.
+
+As we discussed in the [section on risks](https://model-checking.github.io/kani-verifier-blog/2023/02/28/kani-internship-projects-2022-stubbing.html#risks-of-stubbing) stubbing is a powerful tool with few guard rails. In principle you can stub a function with any other function, so long as the type signature matches. Of course any user of stubs is going to be careful to align the behavior of the function being stubbed and the stub. However the whole point of stubbing is that the stub doesn’t behave exactly as the function it replaces, because it needs to be faster. If the approximation is incorrect, it leads to *unsoundness*, meaning that verification can succeed in spite of a bug being present.
+
+In today's blogpost we want to introduce you to a new concept that lets us verify larger programs without sacrificing soundness: function contracts. The idea is very simple: like a stub the contract approximates the behavior of a function in an efficient way, allowing us to scale and verify larger programs. However with a contract the verifier automatically checks that the approximation is sound so we can be confident in the results of our verification.
```

Comment:
This would be a great opening statement for the blog post.I think you should start introducing the concept of modular verification. Talk about breaking down the verification problem, and how that can help you verify a bigger piece of code.

I would then open up with an example and how function contracts can be used to break down the proof and bring the verification time down.

I would add a section on Contracts vs Stubbing at the end of the post.